File: timus/1989.hs
{-# LANGUAGE NPlusKPatterns, BangPatterns, MagicHash #-}

import Control.Applicative
import Control.Monad
import Control.Monad.ST
import Data.Array.ST
import Data.Array.Unboxed (UArray, (!), listArray)
import Data.Bits
import Data.ByteString.Char8 (readInt, words, getContents, lines)
import qualified Data.ByteString.Char8 as BS
import Data.Char
import Data.Int
import Data.Maybe
import Debug.Trace
import GHC.Prim
import GHC.Types
import Prelude hiding (words, lines, getContents)

type Fenwick s = STUArray s Int Int64

makeFenwick n = newArray (0, n-1) 0 :: ST s (Fenwick s)

a `andI#` b = word2Int# (int2Word# a `and#` int2Word# b)
a `orI#` b = word2Int# (int2Word# a `or#` int2Word# b)

update f (i + 1) d = do
	n <- snd <$> getBounds f
	let next (I# i) = I# (i `orI#` (i +# 1#))
	let is = takeWhile (<= n) $ iterate next i
	-- let is = takeWhile (<= n) $ iterate (\i -> i .|. (i + 1)) i
	forM_ is $ \i -> do
		v <- readArray f i
		writeArray f i (v + d)

foldM' :: (Monad m) => (a -> b -> m a) -> a -> [b] -> m a
foldM' _ z [] = return z
foldM' f z (x:xs) = do
  z' <- f z x
  z' `seq` foldM' f z' xs

query f 0 = return 0
query f (i + 1) = do
	-- let is = takeWhile (>= 0) $ iterate (\i -> (i .&. (i+1)) - 1) i
	-- sum <$> mapM (readArray f) is
	let next (I# i) = I# ((i `andI#` (i +# 1#)) -# 1#)
	foldM' (\s i -> (s+) <$> (readArray f i)) 0 $ takeWhile (>= 0) $ iterate next i

data Hashs s = Hashs Int (Fenwick s) (UArray Int Int64) Bool

base :: Int64
base = 13

modulo :: Int64
modulo = 10^9 + 7

a +. b = (a + b) `rem` modulo
a *. b = (a * b) `rem` modulo
a -. b = (a - b + modulo) `rem` modulo

makeHashs s rev = do
		fenwick <- makeFenwick n
		forM_ (zip (BS.unpack s') [1..]) $ \(c, i) -> do
			let d = (fromIntegral (ord c - ord 'a')) *. (coefs ! (i - 1))
			update fenwick i d
		return $ Hashs n fenwick coefs rev	
	where
		n = BS.length s
		coefs = listArray (0, n-1) $ iterate (*. base) 1
		s' = if rev then BS.reverse s else s
		
hashsQuery :: Hashs s -> Int -> Int -> ST s Int64
hashsQuery (Hashs n fenwick coefs rev) l r = do
		v1 <- query fenwick r'
		v2 <- query fenwick (l' - 1)
		let v = v1 -. v2

		return $ if l'+r' <= n+1 then v *. (coefs ! (n+1-r'-l')) else v
	where
		(l', r') = if rev then (n+1-r, n+1-l) else (l, r)

hashsUpdate :: Hashs s -> Int -> Char -> ST s ()
hashsUpdate (Hashs n fenwick coefs rev) i c = do
		v1 <- query fenwick i'
		v2 <- query fenwick (i' - 1)
		let p = v1 -. v2
		update fenwick i' (vv - p)
	where
		i' = if rev then n+1-i else i
		vv = (fromIntegral (ord c - ord 'a')) *. (coefs ! (i' - 1))

readInt' = fst . fromJust . readInt

solve :: BS.ByteString -> [Op] -> [Bool]
solve s ops = runST $ do
	lr <- makeHashs s False
	rl <- makeHashs s True

{-
	vs <- forM ops $ \op -> case op of
		(OpChange i c) -> do
			hashsUpdate lr i c
			hashsUpdate rl i c
			return ""
		(OpQuery i j) -> do
			v1 <- hashsQuery lr i j
			v2 <- hashsQuery rl i j
			return $ if v1 == v2 then "Yes" else "No"
-}

	reverse <$> foldM (\rs op -> case op of
		(OpChange i c) -> do
			hashsUpdate lr i c
			hashsUpdate rl i c
			return rs
		(OpQuery i j) -> do
			v1 <- hashsQuery lr i j
			v2 <- hashsQuery rl i j
			let !r = v1 == v2
			return $ r : rs
		) [] ops

data Op = OpChange !Int !Char | OpQuery !Int !Int deriving (Show)

readOp s =
		if (BS.head a) == 'c'
			then OpChange (readInt' b) (BS.head c)
			else OpQuery (readInt' b) (readInt' c)
	where
		[a, b, c] = words s

main = do
	s <- BS.getLine
	m <- readInt' <$> BS.getLine
	ops <- map readOp <$> replicateM m BS.getLine

	putStrLn . unlines . (map (\v -> if v then "Yes" else "No")) $ solve s ops

File: timus/1989.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
)

type Reader struct {
	*bufio.Scanner
}

func NewReader(r io.Reader) Reader {
	scanner := bufio.NewScanner(r)
	scanner.Split(bufio.ScanWords)
	return Reader{scanner}
}

func (r Reader) String() string {
	b := r.Scan()
	if !b {
		panic(nil)
	}

	return r.Text()
}

func (r Reader) Int() int {
	b := r.Scan()
	if !b {
		panic(nil)
	}

	v, e := strconv.Atoi(r.Text())
	if e != nil {
		panic(e)
	}
	return v
}

func reverse(s string) string {
	b := []byte(s)

	for l, r := 0, len(b)-1; l < r; l, r = l+1, r-1 {
		b[l], b[r] = b[r], b[l]
	}

	return string(b)
}

type Fenwick []int64

func MakeFenwick(n int) Fenwick {
	return Fenwick(make([]int64, n))
}

func (f Fenwick) Update(i int, d int64) {
	for i < len(f) {
		f[i] += d
		i = i | (i + 1)
	}
}

func (f Fenwick) Query(i int) int64 {
	s := int64(0)

	for i >= 0 {
		s += f[i]
		i = i&(i+1) - 1
	}

	return s
}

type Hashs struct {
	fenwick Fenwick
	bytes   []byte
	coefs   []int64
	rev     bool
}

const (
	base   = 13
	modulo = 1000000000 + 7
)

func MakeHashs(s string, rev bool) *Hashs {
	coefs := make([]int64, len(s))
	coefs[0] = 1
	for i := 1; i < len(s); i++ {
		coefs[i] = (coefs[i-1] * base) % modulo
	}

	if rev {
		s = reverse(s)
	}

	fenwick := MakeFenwick(len(s))

	for i := 0; i < len(s); i++ {
		d := int64(s[i]-'a') * coefs[i] % modulo
		fenwick.Update(i, d)
	}

	return &Hashs{
		fenwick,
		[]byte(s),
		coefs,
		rev,
	}

}

func (h Hashs) Update(i int, v byte) {
	if h.rev {
		i = len(h.bytes) - 1 - i
	}

	vv := int64(v-'a') * h.coefs[i] % modulo
	p := (h.fenwick.Query(i) - h.fenwick.Query(i-1) + modulo) % modulo
	h.bytes[i] = v
	h.fenwick.Update(i, vv-p)
}

func (h Hashs) Query(l, r int) int64 {
	if h.rev {
		l, r = len(h.bytes)-1-r, len(h.bytes)-1-l
	}

	v := (h.fenwick.Query(r) - h.fenwick.Query(l-1) + modulo) % modulo

	if l <= len(h.bytes)-1-r {
		v = v * h.coefs[len(h.bytes)-1-r-l] % modulo
	}

	return v
}

func main() {
	in := NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var s string
	fmt.Scan(&s)

	lr := MakeHashs(s, false)
	rl := MakeHashs(s, true)

	n := in.Int()
	for i := 0; i < n; i++ {
		cmd := in.String()

		if cmd[0] == 'c' {
			i, c := in.Int()-1, in.String()

			lr.Update(i, c[0])
			rl.Update(i, c[0])
		} else {
			i, j := in.Int()-1, in.Int()-1

			if lr.Query(i, j) == rl.Query(i, j) {
				out.WriteString("Yes\n")
			} else {
				out.WriteString("No\n")
			}
		}
	}
}
